modules/navier_stokes/src/executioners/SIMPLE.C

```diff
@@ -0,0 +1,1106 @@
+  for (const auto row_i : make_range(matrix.row_start(), matrix.row_stop()))
+    matrix.set(row_i, row_i, diff_diagonal(row_i));
+  matrix.close();
+
+  // Finally, we can create (D*-D) vector which is used for the relaxation of the
+  // right hand side later
+  diff_diagonal.add(-1.0, *original_diagonal);
+  diff_diagonal.close();
+}
+
+void
+SIMPLE::relaxRightHandSide(NumericVector<Number> & rhs,
+                           const NumericVector<Number> & solution,
+                           const NumericVector<Number> & diff_diagonal)
+{
+
+  // We need a working vector here to make sure we don't modify the
+  // (D*-D) vector
+  auto working_vector = diff_diagonal.clone();
+  working_vector->pointwise_mult(solution, *working_vector);
+
+  // The correction to the right hand side is just
+  // (D*-D)*old_solution
+  // For more information see
+  //
+  // Juretic, Franjo. Error analysis in finite volume CFD. Diss.
+  // Imperial College London (University of London), 2005.
+  rhs.add(*working_vector);
+  rhs.close();
+}
+
+Real
+SIMPLE::computeNormalizationFactor(const NumericVector<Number> & solution,
+                                   const SparseMatrix<Number> & mat,
+                                   const NumericVector<Number> & rhs)
+{
+  // This function is based on the description provided here:
+  // @article{greenshields2022notes,
+  // title={Notes on computational fluid dynamics: General principles},
+  // author={Greenshields, Christopher J and Weller, Henry G},
+  // journal={(No Title)},
+  // year={2022}
+  // }
+  // so basically we normalize the residual with the following number:
+  // sum(|Ax-Ax_avg|+|b-Ax_avg|)
+  // where A is the system matrix, b is the system right hand side while x and x_avg are
+  // the solution and average solution vectors
+
+  // We create a vector for Ax_avg and Ax
+  auto A_times_average_solution = solution.zero_clone();
+  auto A_times_solution = solution.zero_clone();
+
+  // Beware, trickery here! To avoid allocating unused vectors, we
+  // first compute Ax_avg using the storage used for Ax, then we
+  // overwrite Ax with the right value
+  *A_times_solution = solution.sum() / solution.size();
+  mat.vector_mult(*A_times_average_solution, *A_times_solution);
+  mat.vector_mult(*A_times_solution, solution);
+
+  // We create Ax-Ax_avg
+  A_times_solution->add(-1.0, *A_times_average_solution);
+  // We create Ax_avg - b (ordering shouldn't matter we will take absolute value soon)
+  A_times_average_solution->add(-1.0, rhs);
+  A_times_solution->abs();
+  A_times_average_solution->abs();
+
+  // Create |Ax-Ax_avg|+|b-Ax_avg|
+  A_times_average_solution->add(*A_times_solution);
+
+  // Since use the l2 norm of the solution vectors in the linear solver, we will
+  // make this consistent and use the l2 norm of the vector
+  // TODO: Would be nice to see if we can do l1 norms in the linear solve
+  return (A_times_average_solution->l2_norm() + 1e-14);
```

Comment:
Some motivation on the choice of 1e-14 might be good. Could also do `TOLERANCE * TOLERANCE` which would be 1e-12

Reply:
I had no motivation, I can do 1e-12 too, just wanted to avoid |r|/0 scenarios.